src/utils/AnalyticsUtils.ts

```diff
@@ -1,6 +1,50 @@
+import { history } from 'coveo.analytics';
 import { IAnalyticsActionCause, analyticsActionCauseList } from '../ui/Analytics/AnalyticsActionListMeta';
+import { findLastIndex } from 'underscore';
+
+export interface IClientInformationProvider {
+  readonly visitorId: string;
+  readonly clientId: string;
+  readonly disabled: boolean;
+}
 
 export class AnalyticsUtils {
+  private static clientInformationProvider: IClientInformationProvider;
+
+  static get visitorId() {
+    return this.analyticsEnabled ? this.clientInformationProvider.visitorId : null;
+  }
+
+  static get clientId() {
+    return this.analyticsEnabled ? this.clientInformationProvider.clientId : null;
+  }
+
+  static get pageId() {
+    const store = new history.HistoryStore();
+    const actions = store.getHistory() as { name: string; value?: string }[];
+    const pageViewActionId = findLastIndex(actions, action => action.name === 'PageView');
+    if (pageViewActionId === -1) {
+      return null;
+    }
+    return actions[pageViewActionId].value;
+  }
+
+  static get location() {
+    return document.location.href;
+  }
+
+  static get referrer() {
+    return document.referrer;
+  }
+
+  private static get analyticsEnabled() {
+    return !!this.clientInformationProvider && !this.clientInformationProvider.disabled;
+  }
+
+  static setClientIdProvider(provider: IClientInformationProvider) {
+    this.clientInformationProvider = provider;
+  }
```

Comment:
This function is added as a bit of a redundant security: it would be possible to add the above getters or their contents elsewhere, but by having them here, some of these getters return `null` when analytics are disabled.

However, I don't find this pattern particularly pretty. Any suggestions?

Reply:
I would suggest avoiding `static`s in favor of `class` instances instead. The provider can be passed to the constructor.

```
const analyticsUtils = new AnalyticsUtils(provider);
```

This would remove the need to check if `clientInformationProvider` is defined (line 41). It guarantees that the provider is defined.

Reply:
That's what I originally wanted, but I'm not sure how I could access that instance from `SearchEndpoint` without making `AnalyticsUtils` static or a singleton.

Reply:
I think a singleton is acceptable in this case because it is initialized with a `provider`, it is read-only. A `static` implementation is okay too, but I think a singleton is a little cleaner.

Reply:
Changing `AnalyticsUtils` into a singleton. However, the check for `clientInformationProvider` is still necessary since some non-identifiable getters are needed even when analytics are disabled.